src/instructlab/config/init.py

```diff
@@ -183,6 +202,59 @@ def init(
     )
 
 
+def hw_auto_detect() -> tuple[str | None, int | None, int | None, _train, bool]:
+    # Third Party
+    import torch
+
+    profiles = render_profiles()
+    gpus = 0
+    total_vram = 0
+    train = _train()
+    chosen_vram = None
+    chosen_profile_name = None
+    previous_vram = None
+    # try nvidia
+    if torch.cuda.is_available():
+        click.echo("Detecting Hardware...")
+        gpus = torch.cuda.device_count()
+        for i in range(gpus):
+            properties = torch.cuda.get_device_properties(i)
+            total_vram += properties.total_memory  # memory in B
+    vram = convert_bytes_to_proper_mag(total_vram)[0]
+    for profile in profiles:
+        # for each profile mapping (vram -> list of profiles)
+        # if the vram we have is < the vram for this group of profiles
+        if vram < profile[0]:
```

Comment:
Why is this organized as a group max and a separate profile max? Would it be simpler / more extensible to just have a lookup table for all known devices, with entries for # of GPUs? For example:
```
card_profiles = {
     'NVIDIA L40S', (2,  2xL40_PROFILE),
     'NVIDIA L40S', (4,  4xL40_PROFILE),
     ...
}
```

And then if you need a generic fallback you can change the device name to GENERIC_NVIDIA if the key is not in card_profiles.

Reply:
I had the groupings almost for a future proofing purpose when we introduce system profiles. The top level "groups" would also have a `Config` Object associated with them for other cmd settings (generation serving etc). This might be a little too early for that though so I can change it to just a single map.


I think `GENERIC_NVIDIA` though will cause more issues than it'll solve for most users. given that for most usecases these profiles are about associating certain settings with a specific amount of vRAM. 

This way too we can give upstream users a better chance at matching their card.

Reply:
lmk if what I changed it to is good for this first version of auto detection!

Reply:
Sorry my reply was a bit too vague. What I meant was that essentially you are implementing GENERIC_NVIDIA, in that you are using a heuristic (vram sizing) to determine which node, vs a known static association. What I meant is that a fixed association with a heuristic as a fallback would have two advantages:

- Simpler implementation, better code clarity
- Less side-effect potential

Using vram is a good choice for an input to a heuristic, but as any heuristic it's only mostly accurate. As an example, a 2x80 H100 and a 4x40 A100 both render the same vram, but we want two different configurations. In the former we need a gpu count of 2 for inferencing (I know this PR isnt addressing that so I am speaking broadly), and in the latter we want 4. We also have architectural generation differences so there may be training features which start to appear in training configurations (e.g. FP8 or other sm_89..sm_90 specific optos).

The advantage of a fixed association comes into play if we need to add a profile variation. Let's say we are close to final release and we need to tweak something just for a specific GPU (say a 4xH200), and so we add some new profiles. The logic will only execute differently only for those specific profiles since it's just a lookup table. In contrast, you might tweak the heuristic algorithm, and now the behavior potentially affects everything, so you could end up with a side-effect breaking another card type unintentionally, and then you update it again and potentially break another and so on. The only away around that is an extensive test that verifies all permutations, wherea a lookup table you can verify by proof and lighter testing.

There is a trade-off though in that you have to update it, for new hardware, but falling back to the heuristic + a generic safe config helps there. For example the generic nvidia configs could assume a low arch sm_70, and then just be sized by vram similar to how you have it. I am traveling today but will comment on your next iteration.

Reply:
@n1hility how about something like a:

`--choose-closest` flag that if specified will match the closest vRAM, otherwise will match based off the GPU name? I feel like this matches both of the usecases we are worried about?

Reply:
> `--choose-closest` flag that if specified will match the closest vRAM, otherwise will match based off the GPU name? I feel like this matches both of the usecases we are worried about?

I might have this wrong but the discovery of such a flag seems unlikely.  Unless you are automating the init call and know that provides a good outcome for your hardware.

Otherwise I would think we have one shot to guess, followed by the user potentially deciding we made a made choice.  If that is the case, seems like we should just focus on making the best guess we can given the hardware.

Reply:
>  If that is the case, seems like we should just focus on making the best guess we can given the hardware.

Is the best guess here based off of vRAM or card name?

The version I have now is more of a "best guess" but will also very clearly match the supported d/s hardware.

This new proposal would only match based off of GPU name and amount, no guessing involved.

Reply:
How about this:

1. check for matching GPU name and Number
2. if none of these match, try to match by vRam

This accounts for
a. the upstream usecase where people just want it to work
b. the concern jason raised with using the wrong heuristic as our first check
c. makes sure all profiles we support get chosen before any type of smart "guessing:

Reply:
I like the above idea of GPU card name followed by matching by vRAM - if neither match I say we fail out with some message like "couldn't autodetect - please select a profile manually"

Reply:
@cdoern +1 having the heuristic fallback is important.